docusaurus/docs/Android/03-ui/06-guides/08-adding-custom-attachments.mdx

```diff
@@ -0,0 +1,169 @@
+            binding.deleteButton.setOnClickListener {
+                onAttachmentCancelled(attachment)
+            }
+        }
+    }
+}
+```
+
+Now you can send a custom attachment of type `date` to the chat. The resulting UI will look like this:
+
+![Sending Custom Attachment](../../assets/sending_custom_attachments.png)
+
+Next, you'll need to build a custom attachment factory to render the item in the message list.
+
+## Rendering Date Attachments
+
+To render a custom attachment in the [MessageListView](../04-message-components/02-message-list.mdx) you'll have to implement `AttachmentFactory` interface:
+
+```kotlin
+interface AttachmentFactory {
+    
+    fun canHandle(message: Message): Boolean
+    
+    fun createViewHolder(
+        message: Message,
+        listeners: MessageListListenerContainer?,
+        parent: ViewGroup,
+    ): InnerAttachmentViewHolder
+}
+```
+
+There are two methods that need to be implemented:
+
+* `canHandle`: Checks whether the factory can handle the given attachments.
+* ```
+* `createViewHolder`: Represents the attachment UI within the `MessageListView`.
+
+Let's see how to create an `AttachmentFactory` that is capable of handling date attachments:
+
+```kotlin
+class DateAttachmentFactory : AttachmentFactory {
+
+    override fun canHandle(message: Message): Boolean {
+        return message.attachments.any { it.type == "date" }
+    }
+
+    override fun createViewHolder(
+        message: Message,
+        listeners: MessageListListenerContainer?,
+        parent: ViewGroup,
+    ): InnerAttachmentViewHolder {
+        return ItemDateAttachmentBinding
+            .inflate(LayoutInflater.from(parent.context), parent, false)
+            .let { DateAttachmentViewHolder(it, listeners) }
+    }
+
+    class DateAttachmentViewHolder(
+        private val binding: ItemDateAttachmentBinding,
+        listeners: MessageListListenerContainer?,
+    ) : InnerAttachmentViewHolder(binding.root) {
+
+        private lateinit var message: Message
+
+        init {
+            binding.dateTextView.setOnClickListener {
+                listeners?.attachmentClickListener?.onAttachmentClick(
+                    message,
+                    message.attachments.first()
+                )
+            }
+            binding.dateTextView.setOnLongClickListener {
+                listeners?.messageLongClickListener?.onMessageLongClick(message)
+                true
+            }
+        }
+
+        override fun onBindViewHolder(message: Message) {
+            this.message = message
+
+            binding.dateTextView.text = message.attachments
+                .first { it.type == "date" }
+                .extraData["payload"]
+                .toString()
+        }
+    }
+}
+```
+
+What's important here is how you're able to fetch the custom attachment data, using `attachment.extraData["payload"]`. You can format the data in any way you want here, which makes our attachments very powerful.
```

Comment:
What's really exciting here is the way the custom attachment data is being fetched, using `attachment.extraData["payload"]`. You can format the data in any way you want here, which makes our attachments very powerful.